### PR TITLE
better handling of size 0 inputs

### LINKED ISSUE #1515: Improve handling of size 0 inputs
## Description
A number of PRs have gone in recently to handle corner cases concerning size 0 inputs that were discovered through the new autodiff test framework, authored in the most part by me.

A couple of comment #1452 ( https://github.com/stan-dev/math/pull/1452#discussion_r352233941 and https://github.com/stan-dev/math/pull/1452#discussion_r352273992) made me look again at these corner cases, and I realised that things could be handled a bit better.

In particular:
- the current checks mishandle matrices of size 0 x N (or N x 0), as we may return an empty square matrix instead of a rectangular one: affects `mdivide_left_ldlt`, `mdivide_right_ldlt` and `matrix_exp_multiply`
- we should check for the input matrix being square before returning anything: affects `matrix_exp_pade` and `scale_matrix_exp_multiply`, for which we currently we return an empty matrix when we should instead throw
- for `trace_inv_quad_form_ldlt` if the input matrix is 0xN we should return 0, but we currently throw

### PR DESCRIPTION
**Pull request by mcol**
## Summary

The following is done:
- `matrix_exp_multiply`, `scale_matrix_exp_multiply` and `matrix_exp_pade`: move `check_square` at the beginning, so that it throws if `A` is not square (currently it returns an empty matrix if any dimension is 0)
- `scale_matrix_exp_multiply`: it returns a rectangular matrix with 0 rows if `A` is empty and `B` has zero rows (currently it returns a square empty matrix)
- `mdivide_left_ldlt`: it returns a rectangular matrix with 0 rows if `A` is empty and `b` has zero rows (currently it returns a square empty matrix)
- `mdivide_right_ldlt`: it returns a rectangular matrix with 0 columns if `A` is empty and `b` has zero columns (currently it returns a square empty matrix)
- `trace_inv_quad_form_ldlt`: use `B.rows()` instead of `B.size()`, so we ensure that matrices are multipliable before returning 0

This also:
- avoids extra work in `check_symmetric` by checking if the size of the square matrix is 0
- same for `check_corr_matrix`, so that for size 0 matrix it returns instead of throwing

Fixes #1515.

## Tests

I've added a bunch of tests that check for all of the above. It's still worth checking for more cases that I might have missed. I'm also not entirely convinced that my mental mapping of prim/rev/mix is complete or correct, so it's possible I've missed something.

## Side Effects

None.

## Checklist

- [X] Math issue #1515

- [X] Copyright holder: Marco Colombo

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [X] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [X] the code is written in idiomatic C++ and changes are documented in the doxygen

- [X] the new changes are tested

### CODE REVIEW

File: stan/math/prim/mat/fun/mdivide_left_ldlt.hpp [hunk 1]
```diff
@@ -22,7 +22,8 @@ template <int R1, int C1, int R2, int C2, typename T1, typename T2>
 inline Eigen::Matrix<return_type_t<T1, T2>, R1, C2> mdivide_left_ldlt(
     const LDLT_factor<T1, R1, C1> &A, const Eigen::Matrix<T2, R2, C2> &b) {
   if (A.cols() == 0 && b.rows() == 0) {
-    return {};
+    Eigen::Matrix<return_type_t<T1, T2>, R1, C2> res(0, b.cols());
+    return res;
```
Comment (rok-cesnovar) [inline stan/math/prim/mat/fun/mdivide_left_ldlt.hpp]: No need to create an intermediate here. Just use the same return style you used in mdivide_right_ldlt.

File: stan/math/rev/mat/fun/matrix_exp_multiply.hpp [hunk 2]
```diff
@@ -25,11 +25,14 @@ template <typename Ta, typename Tb, int Cb>
 inline Eigen::Matrix<typename stan::return_type<Ta, Tb>::type, -1, Cb>
 matrix_exp_multiply(const Eigen::Matrix<Ta, -1, -1>& A,
                     const Eigen::Matrix<Tb, -1, Cb>& B) {
-  if (A.size() == 0 && B.size() == 0)
-    return {};
+  check_square("matrix_exp_multiply", "input matrix", A);
+  if (A.size() == 0 && B.rows() == 0) {
+    Eigen::Matrix<typename stan::return_type_t<Ta, Tb>, -1, Cb> res(0,
+                                                                    B.cols());
+    return res;
```
Comment (rok-cesnovar) [inline stan/math/rev/mat/fun/matrix_exp_multiply.hpp]: same here

File: stan/math/prim/mat/fun/mdivide_left_ldlt.hpp (hunk 1 above)
Comment (mcol) [reply stan/math/prim/mat/fun/mdivide_left_ldlt.hpp]: Thanks!
